### PR TITLE
Feat/extend rewrite middleware

### LINKED ISSUE #571: Rewrite middleware contradicts guidance on rawPathInfo
Hello,

I’ve been using [Network.Wai.Middleware.Rewrite](https://github.com/yesodweb/wai/blob/master/wai-extra/Network/Wai/Middleware/Rewrite.hs) to convert a URL based on compulsory query parameters to one based on paths, in order to place a site I’m upgrading behind a reverse proxy. It’s a bit of a workaround, but the pattern may be useful more generally.

I’d like to add helper functions and some more examples to this module, but I noticed the two functions explicitly update [`rawPathInfo`](https://github.com/yesodweb/wai/blob/master/wai/Network/Wai/Internal.hs#L35) despite that function’s documentation advising against it.

So before I submit the PR, should my functions modify `rawPathInfo` and `rawQueryString` (as the functions in Network.Wai.Middleware.Rewrite), or not (as the documentation for the raw accessors states)?

---

Brief summary of the PR (to follow pending guidance): 
-   Factors out modification of request from generation of middleware (for use with `WPRModifiedRequest` of the `http-reverse-proxy` library)
-   Extend this to use query parameters in addition to path segments
-   Expose new functions to modify requests based on query parameters

This would add extra exports, but not break current exports from the module.

### PR DESCRIPTION
**Pull request by dbaynard**
Fixes #571 

-   The module preserves existing semantics of 'rewrite' and 'rewritePure' but adds warnings.
-   Four new functions are exported.
-   The documentation is expanded, including explanation of the changes.

I wanted to add tests, but couldn’t think of any that didn’t involve deriving `Arbitrary` instances for `Request`.

### CODE REVIEW

**Comment by dbaynard** [comment]
The third commit was because I realised it was probably better to treat query parameters as data rather than text, and handle conversion on a per-use case.

**Comment by dbaynard** [comment]
Think the build failure is because of the import of `Data.Functor.Identity` on GHC < 7.10. I’m not sure how best to handle this.

**Comment by snoyberg** [comment]
This is the problematic bit:

```
Network/Wai/Middleware/Rewrite.hs:280:3: Not in scope: ‘pure’
Network/Wai/Middleware/Rewrite.hs:292:3: Not in scope: ‘pure’
Network/Wai/Middleware/Rewrite.hs:304:41: Not in scope: ‘<$>’
```

You'll need an import from Control.Applicative

**Comment by dbaynard** [comment]
*slaps forehead* forgot about AMP. Thanks! Is the rest ok?

File: wai-extra/Network/Wai/Middleware/Rewrite.hs [hunk 1]
```diff
@@ -1,32 +1,305 @@
...
+import Data.Functor.Identity (Identity(..))
 import qualified Data.Text.Encoding as TE
 import qualified Data.Text as T
 import Network.HTTP.Types as H
 
+-- $howto
+-- This module provides 'Middleware' to rewrite URL paths. It also provides
+-- functions that will convert a 'Request' to a modified 'Request'.
+-- Both operations require a function that takes url parameters and
+-- headers, and returns new url parameters. Parameters are pieces of url
```
Comment (MaxGabriel) [inline wai-extra/Network/Wai/Middleware/Rewrite.hs]: URL should be capitalized

File: wai-extra/Network/Wai/Middleware/Rewrite.hs [hunk 2]
```diff
@@ -1,32 +1,305 @@
...
+-- type of the new function is different the compiler will indicate where
+-- this change must be made.
+--
+-- The 'rewriteRequest' and 'rewriteRequestPure' functions use the new
+-- semantics, too.
 
--- | rewrite based on your own conversion rules
--- Example convert function:
+-- $takeover
+-- Let’s say we want to replace a website written in php with one written
```
Comment (MaxGabriel) [inline wai-extra/Network/Wai/Middleware/Rewrite.hs]: PHP should be capitalized

File: wai-extra/Network/Wai/Middleware/Rewrite.hs [hunk 3]
```diff
@@ -1,32 +1,305 @@
...
+-- > pathFromPhp (pieces, queries) _ = piecesConvert pieces queries
+-- >     where
+-- >         piecesConvert :: [Text] -> H.Query -> PathsAndQueries
+-- >         piecesConvert ["index.php"] qs@(join . lookup "page" -> Just page) =
+-- >             ( ["index", TE.decodeUtf8With TE.lenientDecode page]
+-- >             , delete ("page", pure page) qs
+-- >             )
+-- >         piecesConvert ps qs = (ps, qs)
+--
+-- On the other side, will use 'rewriteRequestPure' to rewrite outgoing
```
Comment (MaxGabriel) [inline wai-extra/Network/Wai/Middleware/Rewrite.hs]: Should be we will use, I think

File: wai-extra/Network/Wai/Middleware/Rewrite.hs [hunk 4]
```diff
@@ -1,32 +1,305 @@
...
+--
+-- /?bar=baz/ would look like
+--
+-- > [("bar", Just "baz")] :: QueryText
+--
+-- Together,
+--
+-- /\/foo?bar=baz/ would look like
+--
+-- > (["foo"],[("bar", Just "baz")]) :: PathsAndQueries
```
Comment (MaxGabriel) [inline wai-extra/Network/Wai/Middleware/Rewrite.hs]: This section was pretty useful for understanding the sample code above. Maybe it should go first, I'm not sure

File: wai-extra/Network/Wai/Middleware/Rewrite.hs [hunk 5]
```diff
@@ -1,32 +1,305 @@
...
+  newReq <- rewriteRequestM convert req
+  let rawPInfo = TE.encodeUtf8 . T.intercalate "/" . pathInfo $ newReq
+  pure newReq { rawPathInfo = rawPInfo }
+{-# WARNING rewriteRequestRawM [
+          "This modifies the 'rawPathInfo' field of a 'Request'."
+        , " This is not recommended behaviour; it is however how"
+        , " this function has worked in the past."
+        , " Use 'rewriteRequestM' instead"] #-}
+
+-- | Produce a function that works on 'PathsandQueries' from one working
```
Comment (MaxGabriel) [inline wai-extra/Network/Wai/Middleware/Rewrite.hs]: "And" in `PathsAndQueries` should be capitalized

File: wai-extra/Network/Wai/Middleware/Rewrite.hs [hunk 6]
```diff
@@ -1,32 +1,305 @@
...
+--   where
+--    piecesConvert [] = ["static", "html", "pages.html"]
+--    piecesConvert route@("pages":_) = "static":"html":route
+-- @
+--
+-- Instead, use this function, supplied to 'rewritePureWithQueries'.
+--
+-- @
+-- staticConvert :: PathsAndQueries -> H.RequestHeaders -> PathsAndQueries
+-- staticConvert pathsAndQueries _ = first (piecesConvert pathsAndQueries)
```
Comment (MaxGabriel) [inline wai-extra/Network/Wai/Middleware/Rewrite.hs]: Is there a difference between this approach and `pathsOnly`?

**Comment by MaxGabriel** [comment]
Really nice documentation, nice work @dbaynard

**Comment by snoyberg** [comment]
There's a build failure on Travis from older GHCs missing `pure` and `<$>`, can you address those? For example, https://travis-ci.org/yesodweb/wai/jobs/152288546.

**Comment by dbaynard** [comment]
Yes, I haven’t had the chance — didn’t realise it was an urgent issue — but I’ll do it now, and I’ll make Max’s changes too.

**Comment by snoyberg** [comment]
Not urgent from my side, just a blocker for merging. I was triaging the emails I received over the past week of vacation so got to this now :)

File: wai-extra/Network/Wai/Middleware/Rewrite.hs (hunk 6 above)
Comment (dbaynard) [inline wai-extra/Network/Wai/Middleware/Rewrite.hs]: Yes. Also there’s a typo 👍 .

To clarify, previous users of the library will have code that looks like `staticConvert'`; this was the example given in the old formulation of the library. That code will still work. However, to migrate to using the new functions they just need to add the single function, `first :: ([Text] -> [Text]) -> PathsAndQueries -> PathsAndQueries` (I’ll add the specialised type to the docs). `pathsOnly` has to handle the `RequestHeaders` parameter too; first does not do this.

**Comment by dbaynard** [comment]
Is it OK to have build warnings from the redundant `Control.Applicative` import for ghc ≥ 7.10, or would you prefer CPP statements?

**Comment by snoyberg** [comment]
It's better to avoid the warnings.

**Comment by dbaynard** [comment]
Woohoo, travis is happy! I can’t wait until there are stackage snapshots covering all the supported compiler versions…

Thank you all for the work you’ve done on this library and others. I figured the least I could do was document the functions I proposed to change.

**Comment by snoyberg** [comment]
Thanks @dbaynard.

@MaxGabriel since you've done a more thorough doc review than I, do you think this is good to go?

**Comment by MaxGabriel** [comment]
I haven't looked at the actual code, but yes the docs are good

**Pull request merged**